### PR TITLE
PERF: Release GIL on some datetime ops

### PR DESCRIPTION
**Pull request by chris-b1**
This is a WIP, but far enough along I thought I'd share and see if the approach was reasonable.  

This releases the GIL on most vectorized field accessors (e.g. `dt.year`) and conversion to and from `Period`.  May be places it could be done - obviously would be nice for parsing, but I'm not sure that's possible.

### CODE REVIEW

**Comment by jreback** [comment]
ohh nice!

can u share some timings?

File: pandas/src/period.pyx [hunk 1]
```diff
@@ -164,10 +165,11 @@ def periodarr_to_dt64arr(ndarray[int64_t] periodarr, int freq):
     out = np.empty(l, dtype='i8')
 
     for i in range(l):
-        if periodarr[i] == iNaT:
-            out[i] = iNaT
-            continue
-        out[i] = period_ordinal_to_dt64(periodarr[i], freq)
+        with nogil:
+            if periodarr[i] == NPY_NAT:
```
Comment (jreback) [inline pandas/src/period.pyx]: move nogil outside the loop

File: pandas/src/period.pyx [hunk 2]
```diff
@@ -245,13 +247,13 @@ def period_ordinal(int y, int m, int d, int h, int min, int s, int us, int ps, i
...
 
-cpdef int64_t period_ordinal_to_dt64(int64_t ordinal, int freq):
+cpdef int64_t period_ordinal_to_dt64(int64_t ordinal, int freq) nogil:
     cdef:
         pandas_datetimestruct dts
         date_info dinfo
         float subsecond_fraction
 
-    if ordinal == iNaT:
+    if ordinal == NPY_NAT:  #TODO: does this break anything?
```
Comment (jreback) [inline pandas/src/period.pyx]: NPY_NAT and iNaT are equivalent

NPY_NAT is just has a c type declared

**Comment by chris-b1** [comment]
Here are some timings - getting a pretty nice speedup.  In single-threaded case things are looking about flat.

```python
In [1]: from pandas.util.testing import test_parallel
In [2]: dti = pd.date_range('1900-1-1', periods=100000)

In [3]: def f():
   ...:     for i in range(4):
   ...:         dti.year
In [4]: @test_parallel(4)
   ...: def g():
   ...:     dti.year

In [8]: %timeit f()
10 loops, best of 3: 25.8 ms per loop

In [9]: %timeit g()
100 loops, best of 3: 7.71 ms per loop
```

File: pandas/tslib.pyx [hunk 3]
```diff
@@ -3982,20 +3991,22 @@ def get_date_field(ndarray[int64_t] dtindex, object field):
...
-        for i in range(count):
-            if dtindex[i] == NPY_NAT: out[i] = -1; continue
+        with nogil:
+            for i in range(count):
+                if dtindex[i] == NPY_NAT: out[i] = -1; continue
 
-            pandas_datetime_to_datetimestruct(dtindex[i], PANDAS_FR_ns, &dts)
-            out[i] = monthrange(dts.year, dts.month)[1]
+                pandas_datetime_to_datetimestruct(dtindex[i], PANDAS_FR_ns, &dts)
+                out[i] = days_per_month_table[is_leapyear(dts.year)][dts.month-1]
```
Comment (jreback) [inline pandas/tslib.pyx]: prob makes sense to define this as a c-function and make it nogil (the days_per_month......)

File: pandas/tslib.pyx [hunk 4]
```diff
@@ -3849,6 +3849,7 @@ def get_time_micros(ndarray[int64_t] dtindex):
 
 
 @cython.wraparound(False)
+@cython.boundscheck(False)
 def get_date_field(ndarray[int64_t] dtindex, object field):
```
Comment (kawochen) [inline pandas/tslib.pyx]: If you declared `field` as `char[:]` instead would you be able to `nogil` the whole thing until `raise`?

Comment (chris-b1) [inline pandas/tslib.pyx]: hmm, tried that out, but cython doesn't seem to take a view of strings like that?  http://stackoverflow.com/questions/28203670/how-to-use-cython-typed-memoryviews-to-accept-strings-from-python

**Comment by jreback** [comment]
@chris-b1 loooks good. can you add a whatsnew note (perf) and squash.

**Comment by chris-b1** [comment]
@jreback - updated

**Comment by jreback** [comment]
thanks!

**Comment by jreback** [comment]
@chris-b1 can you add these (clean then make again to see them)

```
warning: pandas/src/period.pyx:144:24: Use boundscheck(False) for faster access
warning: pandas/src/period.pyx:145:23: Use boundscheck(False) for faster access
warning: pandas/src/period.pyx:147:55: Use boundscheck(False) for faster access
warning: pandas/src/period.pyx:148:19: Use boundscheck(False) for faster access
warning: pandas/src/period.pyx:169:24: Use boundscheck(False) for faster access
warning: pandas/src/period.pyx:170:19: Use boundscheck(False) for faster access
warning: pandas/src/period.pyx:172:15: Use boundscheck(False) for faster access
warning: pandas/src/period.pyx:172:53: Use boundscheck(False) for faster access
building 'pandas._period' extension
```

**Pull request merged**